Modules/Loadable/Markups/MRML/vtkMRMLMarkupsCurveNode.cxx

```diff
@@ -183,6 +194,182 @@ double vtkMRMLMarkupsCurveNode::GetCurveLengthBetweenStartEndPointsWorld(vtkIdTy
+	  vtkDataArray* normalVectorArray = vtkArrayDownCast<vtkDataArray>(normalPolydata->GetPointData()->GetNormals());
+	  if (!normalVectorArray)
+	    {
+		vtkGenericWarningMacro("vtkMRMLMarkupsCurveNode::ResamplePoints failed: Unable to find or calculate normals");
+		return;
+		}
+	  }
+
+	double interpolatedPoint[3] = { 0.0 };
+	double projectedPoint[3] = { 0.0 };
+	double segmentStartPoint[3] = { 0.0 };
+	double segmentEndPoint[3] = { 0.0 };
+	vtkNew<vtkPoints> pointNormalArray;
+	//Project each mesh point onto the most external model surface point along a normal vector ray. The normal vector
+	//for each resampled point is estimated from normal vectors at the closest control points on either side.
+	//If there is no intersection between the model and normal vector ray, use the closest model point.
+	for (vtkIdType controlPointIndex = 0; controlPointIndex < interpolatedPoints->GetNumberOfPoints(); controlPointIndex++)
+	  {
+	  interpolatedPoints->GetPoint(controlPointIndex, interpolatedPoint);
+	  vtkIdType segmentStartIndex = vtkMRMLMarkupsNode::GetClosestControlPointIndexToPositionWorld(interpolatedPoint);
+	  originalControlPoints->GetPoint(segmentStartIndex, segmentStartPoint);				// get nearest control point
+	  this->GetClosestPointPositionAlongCurveWorld(interpolatedPoint, projectedPoint);	// get projection of point to curve
+	  if (segmentStartIndex == 0)
+		{
+		originalControlPoints->GetPoint(segmentStartIndex + 1, segmentEndPoint);
+		}
+	  else if (segmentStartIndex == (originalControlPoints->GetNumberOfPoints() - 1))
+	    {
+	    originalControlPoints->GetPoint(segmentStartIndex - 1, segmentEndPoint);
+	    }
+	  else
+		{
+		double* segmentEndPoint1 = originalControlPoints->GetPoint(segmentStartIndex - 1);
+		double dist1 = vtkMath::Distance2BetweenPoints(segmentEndPoint1, projectedPoint);
+		double* segmentEndPoint2 = originalControlPoints->GetPoint(segmentStartIndex + 1);
+		double dist2 = vtkMath::Distance2BetweenPoints(segmentEndPoint2, projectedPoint);
+		if ((dist1 < dist2) && dist1 < vtkMath::Distance2BetweenPoints(segmentEndPoint1, segmentStartPoint))
+		  {
+		  segmentEndPoint[0] = segmentEndPoint1[0];
+		  segmentEndPoint[1] = segmentEndPoint1[1];
+		  segmentEndPoint[2] = segmentEndPoint1[2];
+		  }
+		else
+		  {
+		  segmentEndPoint[0] = segmentEndPoint2[0];
+		  segmentEndPoint[1] = segmentEndPoint2[1];
+		  segmentEndPoint[2] = segmentEndPoint2[2];
+		  }
+		}
+	  // estimate normal vector from control points on either side of projected point
+	  vtkIdType pointIdStart = pointLocator->FindClosestPoint(segmentStartPoint);
+	  double startNormal[3] = { 0.0 };
+	  normalVectorArray->GetTuple(pointIdStart, startNormal);
+	  vtkIdType pointIdEnd = pointLocator->FindClosestPoint(segmentEndPoint);
+	  double endNormal[3] = { 0.0 };
+	  normalVectorArray->GetTuple(pointIdStart, endNormal);
+	  double rayDirection[3] = { 0.0 };
+	  rayDirection[0] = (startNormal[0] + endNormal[0]) / 2.0;
```

Comment:
since we compute distance to start and end points anyway, why don't we use those to weight the average (instead of always using 0.5 weight for both)?